Static middleware

tmattio opened the pull request:
This PR implements a static serve middleware in `opium_kernel`. The implementation is based on the previous `Opium.Static_serve.m` middleware, but takes a function `read : string -> string option Lwt.t`.

This allows using the middleware with different `read` implementation (e.g. MirageOS's filesystem, ocaml-crunch, etc.)

I refactored the existing `Static_serve.m` middleware to use `Opium_kernel.Middleware.static`. The API is the same as the previous implementation.

anuragsoni commented on opium_kernel/src/opium_kernel.mli:
```diff
@@ -213,6 +213,24 @@ module Server_connection : sig
+    -> Rock.Handler.t
+
+  val m
+    :  read:(string -> Rock.Body.t option Lwt.t)
```
I think it might be nicer if we have `read` have the type `string -> (Rock.Body.t, [Status.client_error | Status.server_error]) Lwt_result.t`.

tmattio replied:
Done!

anuragsoni commented on opium_kernel/src/opium_kernel.mli:
```diff
@@ -213,6 +213,24 @@ module Server_connection : sig
 end
 
+module Static : sig
+  val serve
```
not sure why we need to expose this in the mli

tmattio replied:
The static module, as implemented here, can be used in two ways:

- As a middleware, that will read the file with the given filename (the function `m`)
- As a `Rock.Handler` (the function `serve`)

I had a use case for the second point, where I wanted to expose a file with a custom URL.

tmattio commented:
Hi @anuragsoni!

Sorry for the delay. I updated the signature of the read function as you suggested and I addressed the other comments. Let me know if there's any other changes you'd like me to make.

Thanks!

tmattio commented on opium_kernel/src/static.ml:
```diff
@@ -0,0 +1,111 @@
+    Lwt.return (resp, body)
+;;
+
+let _serve ~read ?mime_type ?etag_of_fname ?(headers = Httpaf.Headers.empty) fname req =
```
Done, there's only one function serve now.

tmattio commented on opium_kernel/src/opium_kernel.mli:
```diff
@@ -228,8 +246,16 @@ module Middleware : sig
     -> unit
     -> Rock.Middleware.t
+
+  val static
```
I removed `Static.m` from the `mli`

tmattio commented on opium_kernel/src/opium_kernel.mli:
```diff
@@ -194,15 +194,15 @@ module Router : sig
 
   val empty : 'action t
-  val add : 'a t -> route:Route.t -> meth:Method.t -> action:'a -> 'a t
+  val add : 'a t -> route:Route.t -> meth:Rock.Method.t -> action:'a -> 'a t
```
Reverted.

rgrinberg commented:
Looks good to me. Can you rebase?

tmattio commented:
Thanks for the reviews @anuragsoni and @rgrinberg! I've rebased and squashed. Merging now since it looks ok for the both of you 🙂
